🚧 WIP: Reworked extending SEO extension & new features

jaspervriends opened the pull request:
I'm currently working on improving the way how other extensions can extend on the SEO extension.

New is the `V17Development\FlarumSeo\Extend\SEO` class that can be used as driver manager inside another extensions `extend.php`. 

Using `->addDriver('name', ClassName::class)`, you can add new classes that can manage meta tags and JSON for specific routes.

The `PageListener` is still in use, however, the code in the `determine()` function has been changed and now loops through the returned drivers of the `getDrivers($routeName)` function inside `PageManager`.

An `extend.php` from another extension can now look like:
```
use V17Development\FlarumSeo\Extend\SEO;

return [
    (new SEO())
        ->addDriver('index', IndexPage::class)
        ->addDriver('tags', TagPage::class)
];
```

I personally thought it would be handy to filter on extension dependencies (if any is required) and have a function that returns an array of routes where the class should be triggered. A good example of both functionality is the `TagPage`

```
interface PageDriverInterface
{
    /**
     * A list of Flarum extension IDs for extensions that should be enabled
     */
    public function extensionDependencies(): array;

    /**
     * A list of route names that will be handled
     *
     * Empty array if handles for all routes
     */
    public function handleRoutes(): array;

    /**
     * Handle page SEO
     */
    public function handle(ServerRequestInterface $request, SeoProperties $seo);
}
```

The old class `Extend` has been renamed to `SeoProperties` and will be send to the `handle()` function. The static functions have been updated to normal functions.

luceos commented on src/Extend/SEO.php:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @param string $name Driver name
+     */
+    public function unregisterDriver(string $name)
```
if `add` is the opposite then this method ought to be named `delete` or `remove` not `unregister`

luceos commented on src/Extend/SEO.php:
```diff
@@ -0,0 +1,57 @@
+    public function unregisterDriver(string $name)
+    {
+        // Only unregister if driver existed
+        if (isset($this->drivers[$name])) {
+            unset($this->drivers[$name]);
+        }
```
```suggestion
        Arr::forget($this->drivers, $name);
```

Don't forget to import the Arr helper.

luceos commented on src/Listeners/PageListener.php:
```diff
@@ -119,48 +108,12 @@ private function determine($serverRequest)
-            }
+        // Handle through drivers
+        foreach ($this->pageManager->getDrivers($routeName) as $driverClass) {
+            resolve($driverClass::class)->handle($serverRequest, $seoPropertiesExtender);
```
In the SEO extender method `extend` you are registering the drivers and resolving them through the container already. Using `resolve` again here is unnecessary, they are already resolved. If this is not the case, you have an error elsewhere.

luceos commented on src/Listeners/PageListener.php:
```diff
@@ -199,18 +152,15 @@ private function setSiteTags()
         // Set image
-        if($applicationSeoSocialMediaImage !== null)
-        {
+        if ($applicationSeoSocialMediaImage !== null) {
```
Formatting changes should be made outside of a (massive) feature PR.

Ignore this comment once read ;)

luceos commented on src/Page/PageManager.php:
```diff
@@ -0,0 +1,53 @@
+    public function getDrivers(string $routeName = null)
+    {
+        // Filter available drivers
+        return array_filter($this->drivers, function (PageDriverInterface $driver) use ($routeName) {
+            // Filter drivers that require extensions to be enabled
+            foreach ($driver->extensionDependencies() as $extensionId) {
+                if (!$this->extensionManager->isEnabled($extensionId)) {
+                    return false;
+                }
+            }
+
+            // Check if driver should be handled
+            if (!is_null($routeName) && count($driver->handleRoutes()) !== 0 && !in_array($routeName, $driver->handleRoutes())) {
+                return false;
+            }
+
+            return true;
+        });
```
This cannot be inside this one method, you can see the code smell from the level of indentation as well as the need for comments. Clean, understandable code is best written by keeping methods as small and focused on one responsibility.

Split this code up so that you have one method: `getActiveDrivers` which returns a `Collection` filtered by drivers of which the extension dependencies are enabled.

Then this method can `filter` based on the route:

```suggestion
        return $this->getActiveDrivers()
            ->filter(fn (PageDriverInterface $driver) => $routeName === null || in_array($routeName, $driver->handleRoutes() ?? []))
            ->toArray();
```

Untested ofcourse ;)

davwheat replied:
Arrow functions aren't supported in PHP 7.3, I don't think?

jaspervriends replied:
Ah, I thought the minimum PHP version was 8 and higher. I guess I've mistaken it with Symfony 's minimum requirements instead 👀 

Will fix it 👍

luceos replied:
@davwheat 7.3 isn't even one of the supported versions anymore. This means any critical vulnerability will not be patched 🎉 

https://www.php.net/supported-versions.php

davwheat replied:
That's fair. I just assumed we'd still want to support it as Flarum still does.

luceos replied:
I think core is stuck to 7.3 because we can't just increase it between minor versions. I think 2.0 and L9/10 will bring a minimum of 8+ anyway.

luceos commented on src/Page/TagPage.php:
```diff
@@ -0,0 +1,93 @@
+
+                // Canonical url
+                ->setCanonicalUrl('/t/' . $tag->slug);
+        } catch (\Exception $e) {
```
I don't think catching exceptions should be the responsibility of the `handle` method. This should be processed (**and never fully silenced**) from the PageManager class that triggers the `handle` method.

PS silencing root Exception classes will cause you a ton of pain in maintenance, you will never get decent error reporting from users.

If you expect exceptions, capture their exact classes and handle them accordingly. Not sure? Log them to the LoggerInterface. I personally prefer triggering more exceptions than less, the sooner someone sees an issue, the better it is to tackle the underlying issues.

luceos reviewed (commented):
A ton has improved on this PR, well done.

In the future don't mingle formatting changes with actual feature/refactoring changes, it makes reviewing very tiring.

I left a few recommendations.

jaspervriends commented:
### 2024 update:

Meta tags are now partly autogenerated from the `SeoMeta` model. When a discussion has been created, updated or deleted, a `SeoMeta` will be triggered and updated to the newest version.

If someone enters a page where the `SeoMeta` wasn't generated yet, it will auto-generate it.

The discussion page would handle `SeoMeta` models like this: https://github.com/v17development/flarum-seo/blob/f931080ac1ce7bb9c5f19a7c1129541283d44c93/src/Page/DiscussionPage.php#L105

An example listener, that also includes code to update the `SeoMeta`, looks like this: https://github.com/v17development/flarum-seo/blob/f931080ac1ce7bb9c5f19a7c1129541283d44c93/src/Listeners/SeoMetaListeners/DiscussionListener.php#L23

Also, I created a `findModelOrCreate` that allows a function to be executed. Would the be a go or a no-go? https://github.com/v17development/flarum-seo/blob/f931080ac1ce7bb9c5f19a7c1129541283d44c93/src/SeoMeta/SeoMeta.php#L82

Any feedback would be highly apreciated!

iamdarkle commented:
I've been looking over the PR a bit, and since I encountered the 'Missing field "mainEntity"' error on the profile pages, I thought it would make a good suggestion. I noticed you've implemented `mainEntity` in `src/Page/DiscussionBestAnswerPage.php`. It might be a good idea to also add the schema to `src/Page/ProfilePage.php`, which has been considered for some time now (https://developers.google.com/search/docs/appearance/structured-data/profile-page). Thanks for the massive effort!

jaspervriends commented:
Thanks for checking @iamdarkle , I just noticed the #90 PR as well, will merge that into this. Does the rest of the PR look good to you, or do you perhaps have some feedback on improvements?

jaspervriends commented:
Added SEO meta-object editor

![image](https://github.com/user-attachments/assets/1e756192-95a6-4dbd-ac53-f4f01b1ec2d7)

iamdarkle commented:
> Thanks for checking @iamdarkle , I just noticed the #90 PR as well, will merge that into this. Does the rest of the PR look good to you, or do you perhaps have some feedback on improvements?

I didn't see this notification at the time, in my humble opinion (I'm not a SEO expert 😅) it looks great, a ton of improvements.

> ![image](https://private-user-images.githubusercontent.com/4417659/371896846-1e756192-95a6-4dbd-ac53-f4f01b1ec2d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3NDY1ODYsIm5iZiI6MTcyNzc0NjI4NiwicGF0aCI6Ii80NDE3NjU5LzM3MTg5Njg0Ni0xZTc1NjE5Mi05NWE2LTRkYmQtYWM1My1mNGYwMWIxZWMyZDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDFUMDEzMTI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGRmYjIwMDFjMDI3M2RkNTU0OTVjMTE4NWUyYWMxYjAxODZlOTUyZGZmNjY0NjdlMzgwZDI0MmIzNTgyOWM2NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.TkGz543N9Q-ZKtykyD5HIOsK6eOmW88aESUxwex-dUM)

This UI looks awesome Jasper!
